packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts

```diff
@@ -166,16 +168,71 @@ export async function createGraphqlEngineBundle(
   })
 
   return new Promise((resolve, reject) => {
-    compiler.run((err, stats) => {
-      compiler.close(closeErr => {
-        if (err) {
-          return reject(err)
+    compiler.run((err, stats): void => {
+      function getResourcePath(
+        webpackModule?: Module | NormalModule | ConcatenatedModule | null
+      ): string | undefined {
+        if (webpackModule && !(webpackModule instanceof ConcatenatedModule)) {
+          return (webpackModule as NormalModule).resource
+        }
+
+        if (webpackModule?.modules) {
+          // ConcatenatedModule is a collection of modules so we have to go deeper to actually get a path,
+          // at this point we won't know which one so we just grab first module here
+          const [firstSubModule] = webpackModule.modules
+          return getResourcePath(firstSubModule)
         }
-        if (closeErr) {
-          return reject(closeErr)
+
+        return undefined
+      }
+
+      function iterateModules(
+        webpackModules: Set<Module>,
+        compilation: Compilation
+      ): void {
+        for (const webpackModule of webpackModules) {
+          if (webpackModule instanceof ConcatenatedModule) {
+            iterateModules(
+              (webpackModule as ConcatenatedModule).modules,
+              compilation
+            )
+          } else {
+            const resourcePath = getResourcePath(webpackModule)
+            if (resourcePath?.includes(`ts-node`)) {
+              const importedBy = getResourcePath(
+                compilation.moduleGraph.getIssuer(webpackModule)
+              )
+              const structuredError = {
+                id: `98011`,
+                context: {
+                  package: `ts-node`,
+                  importedBy,
+                  advisory: `Gatsby is supporting TypeScript natively (see https://www.gatsbyjs.com/docs/how-to/custom-configuration/typescript/). "ts-node" might not be needed anymore at all, consider removing it.`,
+                },
+              }
+              throw structuredError
+            }
```

Comment:
This is main course of this change - adds a way to inspect what modules were used for engine and if there are any known problematic ones (`ts-node` in this case), we can show more useful information to user rather than often times very cryptic errors.